ui/tempo-plugin/src/plugins/tempo-trace-query/query-editor-model.ts

```diff
@@ -0,0 +1,58 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useState } from 'react';
+import { produce } from 'immer';
+import { OptionsEditorProps } from '@perses-dev/plugin-system';
+import { TempoTraceQuerySpec } from '../../model/trace-query-model';
+
+export type TraceQueryEditorProps = OptionsEditorProps<TempoTraceQuerySpec>;
+
+/**
+ * A hook for managing the `query` state in PrometheusTimeSeriesQuerySpec. Returns the `query` value, along with
+ * `onChange` and `onBlur` event handlers to the input. Keeps a local copy of the user's input and only syncs those
+ * changes with the overall spec value once the input is blurred to prevent re-running queries in the panel's preview
+ * every time the user types.
+ */
+export function useQueryState(props: TraceQueryEditorProps) {
+  const { onChange, value } = props;
+
+  // Local copy of the query's value
+  const [query, setQuery] = useState(value.query);
+
+  // This is basically "getDerivedStateFromProps" to make sure if spec's value changes external to this component,
+  // we render with the latest value
+  const [lastSyncedQuery, setLastSyncedQuery] = useState(value.query);
+  if (value.query !== lastSyncedQuery) {
+    setQuery(value.query);
+    setLastSyncedQuery(value.query);
+  }
+
+  // Update our local state's copy as the user types
+  const handleQueryChange = (e: string) => {
+    setQuery(e);
+  };
+
+  // Propagate changes to the query's value when the input is blurred to avoid constantly re-running queries in the
+  // PanelPreview
+  const handleQueryBlur = () => {
```

Comment:
Side note: sometimes this onBlur logic annoys me in our Prometheus editor and I just want an explicit 'Run queries' button 😂

Reply:
I agree. This could be generated as a separate issue. Updating the panel to have an explicit 'Run queries' button might also be more intuitive for the user.